File: src/main/java/org/exercises/RandIntProducer.java
package org.exercises;

import org.apache.kafka.clients.admin.AdminClient;
import org.apache.kafka.clients.admin.AdminClientConfig;
import org.apache.kafka.clients.admin.CreateTopicsResult;
import org.apache.kafka.clients.admin.NewTopic;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.errors.TopicExistsException;
import org.apache.kafka.common.serialization.IntegerSerializer;
import org.apache.kafka.common.serialization.StringSerializer;

import java.util.Collections;
import java.util.Properties;
import java.util.Random;
import java.util.concurrent.ExecutionException;

public class RandIntProducer {
    public static void main(String[] args) {
        String server = "localhost:9092";
        String topic = "rand-int-topic";

        Properties props = new Properties();
        props.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");

        try (final AdminClient adminClient = AdminClient.create(props)) {
            try {
                NewTopic newTopic = new NewTopic(topic, 3, (short)1);
                final CreateTopicsResult createTopicsResult = adminClient.createTopics(Collections.singleton(newTopic));
                createTopicsResult.values().get(topic).get();
            } catch (InterruptedException | ExecutionException e) {
                if (!(e.getCause() instanceof TopicExistsException))
                    throw new RuntimeException(e.getMessage(), e);
            }
        }

        Properties properties = new Properties();
        properties.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, server);
        properties.setProperty(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        properties.setProperty(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, IntegerSerializer.class.getName());
        properties.setProperty(ProducerConfig.PARTITIONER_ADPATIVE_PARTITIONING_ENABLE_CONFIG, "false");

        KafkaProducer<String, Integer> producer = new KafkaProducer<>(properties);
        Random random = new Random();
        String type;

        while (true) {
            try {
                int number = random.nextInt(1000);
                if (number % 2 == 0)
                    type = "even";
                else
                    type = "odd";
                ProducerRecord<String, Integer> producerRecord =
                        new ProducerRecord<>(topic, type, number);
                producer.send(producerRecord);

                Thread.sleep(100);
            } catch (InterruptedException e) {
                producer.flush();
                producer.close();
                throw new RuntimeException(e);
            }
        }
    }
}

File: src/main/python/kafka/RandIntProducer.py
from random import randint
from time import sleep

import kafka.errors
from kafka.admin import KafkaAdminClient, NewTopic
from kafka.producer import KafkaProducer
from confluent_kafka.serialization import IntegerSerializer, StringSerializer

host = "localhost:9092"
topic = "kafka-apex-python-rand-int"

admin_client = KafkaAdminClient(
    bootstrap_servers=host
)

try:
    topic_list = []
    topic_list.append(NewTopic(name=topic, num_partitions=3, replication_factor=1))
    admin_client.create_topics(new_topics=topic_list, validate_only=False)
except kafka.errors.TopicAlreadyExistsError as e:
    pass

kafka_producer = KafkaProducer(
    bootstrap_servers = host,
    key_serializer = StringSerializer(),
    value_serializer = IntegerSerializer()
)

integer_type = ""

try:
    while(True):
        rand_int = int(randint(0, 1000))
        if rand_int % 2 == 0:
            integer_type = "even"
        else:
            integer_type = "odd"

        kafka_producer.send(topic = topic, key = integer_type, value = rand_int)
        sleep(1)
except (InterruptedError, KeyboardInterrupt) as e:
    kafka_producer.flush()
    kafka_producer.close()
    raise Exception(e)